docs/src/interface-definition.md

```diff
@@ -0,0 +1,167 @@
+- `AtomsCalculators.forces!` for non-allocating force calculation
+- `AtomsCalculators.virial` for [virial](https://en.wikipedia.org/wiki/Virial_stress) calculation
+
+You do not need to implement all of these. To implement force calculation you only need to implement either allocating or non-allocating force call - see below for more details on how to implement force calculation.
+
+Each call has two inputs: `AtomsBase.AbstractSystem` compatible structure and a `calculator` that incudes details of the calculation method. Additionally keywords can be give. These can be ignored, but they need to be present in the function definition.
+
+- First input is `AtomsBase.AbstractSystem` compatible structure
+- Second input is `calculator` structure
+- Method has to accept keyword arguments (they can be ignored)
+- Non-allocating force call `force!` has an AbstractVector as the first input, to which the evaluated force values are stored (look for more details below)
+
+Outputs for the functions need to have following properties
+
+- Energy is a subtype of `Number` that has a unit with dimensions of energy (mass * length^2 / time^2)
+- Force output is a subtype of `AbstractVector` with element type also a subtype of AbstractVector (length 3 in 3D) and unit with dimensions of force (mass * length / time^2). With additional property that it can be reinterpret as a matrix
+- Virial is a square matrix (3x3 in 3D) that has units of energy times length
+
+
+## Example implementations
+
+Example `potential_energy` implementation
+
+```julia
+function AtomsCalculators.potential_energy(system, calculator::MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition here
+    return 0.0u"eV"
+end
+```
+
+Example `virial` implementation
+
+```julia
+function AtomsCalculators.virial(system, calculator::MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition here
+    return zeros(3,3) * u"eV*Å"
+end
+```
+
+### Implementing forces call
+
+There are two optional implementations for force call allocating and non-allocating. The reason for this is that for very fast potentials allocation has noticeable effect on total evaluation time. So, in order to reduce the evaluation time, there is non-allocating option. On the other hand some expensive methods, like those in quantum chemistry, always allocate output data.
+
+To make implementation easy for everyone, we made it so that you need only to define either allocating or non-allocating force call. The other call is generated is then generated with a macro `AtomsCalculators.@generate_complement`.
+
+Example
+
+```julia
+struct MyType
+end
+
+AtomsCalculators.@generate_complement function AtomsCalculators.forces(system, calculator::Main.MyType; kwargs...)
```

Comment:
this is very interesting. very elegant if it works as expected.

Reply:
Yes it works - it the tests at least.